### PR TITLE
sources: make submodule fetching configurable

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `./runtests.sh static`?
- [X] Have you successfully run `./runtests.sh tests/unit`?

-----
Alternative solution from https://github.com/snapcore/snapcraft/pull/3623

https://bugs.launchpad.net/snapcraft/+bug/1957767

Configure which submodules to fetch from the source tree in snapcraft.yaml with `source-submodules: <list-of-submodules>`. 

#### Defined
If `source-submodules` is defined, only the listed submodules are fetched.
```
parts:
  git-test:
    plugin: dump
    source-type: git
    source: git@github.com...
    source-submodules:
      - submodule_1
      - dir1/submodule_2
```

#### Defined and Empty
If `source-submodules` in defined and empty, no submodules are fetched.
```
parts:
  git-test:
    plugin: dump
    source-type: git
    source: git@github.com...
    source-submodules: []
```

#### Not Defined
If `source-submodules` is not defined, all submodules are fetched (default behavior).

#### Note
If a submodule is fetched, any submodules of that submodules are fetched too.  I don't think the behavior needs to be any more nuanced (i.e. no need to define which submodules to fetch recursively and which to fetch non-recursively).

### CODE REVIEW

File: tests/spread/general/sources/snaps/git-submodules/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+description: Make sure source-submodules works
+confinement: strict
+
+parts:
+    git:
+        plugin: dump
+        source: https://github.com/snapcore/core18
+        source-type: git
+        source-submodules:
+          - submodule_1
```
Comment (sergiusens) [inline tests/spread/general/sources/snaps/git-submodules/snapcraft.yaml:14]: these submodules exist?

File: tests/spread/general/sources/snaps/git-submodules/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+name: git-recurse-submodules
+base: core20
```
Comment (sergiusens) [inline tests/spread/general/sources/snaps/git-submodules/snapcraft.yaml:2]: if this general suite runs on core20 this is fine, my brain doesn't feel comfortable with cloning core18 though

**Comment by sergiusens** [comment]
Looks like state tracking tests have failed

File: tests/spread/general/sources/snaps/git-submodules/snapcraft.yaml (hunk 1 above)
Comment (mr-cal) [reply tests/spread/general/sources/snaps/git-submodules/snapcraft.yaml:14]: No.  Using `--recurse-submodules=<non-existant-directory>` with `git clone` doesn't produce any errors.  The lack of error or warning surprises me, but from what I gather, it's intentional. 

This test only proves that the parameters are parsed and the `git clone` command runs.

Perhaps a slightly better test would be to choose a different repo with a submodule.  Unfortunately I don't see any in snapcore.  The multipass repo has some submodules.  Alternatively I could clone a repo from an outside project.  Think it's worth the trouble?

File: tests/spread/general/sources/snaps/git-submodules/snapcraft.yaml (hunk 2 above)
Comment (mr-cal) [reply tests/spread/general/sources/snaps/git-submodules/snapcraft.yaml:2]: Agreed.

**Comment by mr-cal** [comment]
Sorry for 2 force pushes, I made some mistakes.

**Review by cmatsuoka — APPROVED** [review]
LGTM. We'll need to bring this to legacy snapcraft in the snapcraft/7.0 branch and to craft-parts.

**Pull request merged**